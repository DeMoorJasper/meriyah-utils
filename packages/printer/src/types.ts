import type { ESTree } from "meriyah";

import type { State } from "./core";

export type ExpressionPrecedenceType = Partial<{
  [type in ESTree.Node["type"]]: number;
}>;

export type CodeGeneratorType = Partial<{
  [type in ESTree.Node["type"]]: (node: ESTree.Node, state: State) => void;
}>;

/**
 * Code generator options.
 */
export interface IOptions {
  /**
   * String to use for indentation, defaults to `"␣␣"`.
   */
  indent?: string;
  /**
   * String to use for line endings, defaults to `"\n"`.
   */
  lineEnd?: string;
  /**
   * Indent level to start from, defaults to `0`.
   */
  startingIndentLevel?: number;
  /**
   * Generate comments if `true`, defaults to `false`.
   */
  comments?: boolean;
  /**
   * Custom code generator logic.
   */
  generator?: CodeGeneratorType;

  /**
   * Precedence of expressions
   */
  expressionsPrecedence?: ExpressionPrecedenceType;
}
