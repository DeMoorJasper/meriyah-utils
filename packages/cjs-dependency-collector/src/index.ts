import type { ESTree } from "meriyah";

import { simpleWalk } from "@meriyah-utils/walker";
import { NodeTypes } from "@meriyah-utils/types";

export function collectDependencies(program: ESTree.Program) {
  const deps: Set<string> = new Set();
  simpleWalk(program, (node) => {
    if (
      node.type === NodeTypes.CallExpression &&
      node.callee.name === "require"
    ) {
      // @ts-ignore
      if (node.arguments.length && node.arguments[0].value) {
        // @ts-ignore
        deps.add(node.arguments[0].value);
      }

      return true;
    }
  });
  return Array.from(deps);
}
