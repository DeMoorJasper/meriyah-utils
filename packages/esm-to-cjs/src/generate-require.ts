import { ESTree } from "meriyah";
import { NodeTypes as n } from "@meriyah-utils/types";

/**
 * Generates `var $varName = require('$requirePath');
 */
export function generateRequireStatement(varName: string, requirePath: string) {
  return {
    type: n.VariableDeclaration,
    declarations: [
      {
        type: n.VariableDeclarator,
        id: {
          type: n.Identifier,
          name: varName,
        },
        init: {
          type: n.CallExpression,
          callee: {
            type: n.Identifier,
            name: "require",
          },
          arguments: [
            {
              type: n.Literal,
              value: requirePath,
            },
          ],
        },
      },
    ],
    kind: "var" as "var",
  };
}

export function generateInteropRequire() {
  return {
    type: n.FunctionDeclaration,
    params: [
      {
        type: n.Identifier,
        name: "obj" as "obj",
      },
    ],
    body: {
      type: n.BlockStatement,
      body: [
        {
          type: n.ReturnStatement,
          argument: {
            type: n.ConditionalExpression,
            test: {
              type: n.LogicalExpression,
              left: {
                type: n.Identifier,
                name: "obj",
              },
              right: {
                type: n.MemberExpression,
                object: {
                  type: n.Identifier,
                  name: "obj",
                },
                computed: false,
                property: {
                  type: n.Identifier,
                  name: "__esModule",
                },
              },
              operator: "&&",
            },
            consequent: {
              type: n.Identifier,
              name: "obj",
            },
            alternate: {
              type: n.ObjectExpression,
              properties: [
                {
                  type: n.Property,
                  key: {
                    type: n.Identifier,
                    name: "default",
                  },
                  value: {
                    type: n.Identifier,
                    name: "obj",
                  },
                  kind: "init" as "init",
                  computed: false,
                  method: false,
                  shorthand: false,
                },
              ],
            },
          },
        },
      ],
    },
    async: false,
    generator: false,
    id: {
      type: n.Identifier,
      name: "$_csb__interopRequireDefault",
    },
  };
}

export function generateInteropRequireExpression(
  argument: ESTree.Expression,
  localName: string
) {
  return {
    type: n.VariableDeclaration,
    kind: "var" as "var",
    declarations: [
      {
        type: n.VariableDeclarator,
        init: {
          type: n.CallExpression,
          callee: {
            type: n.Identifier,
            name: "$_csb__interopRequireDefault",
          },
          arguments: [argument],
        },
        id: {
          type: n.Identifier,
          name: localName,
        },
      },
    ],
  };
}
