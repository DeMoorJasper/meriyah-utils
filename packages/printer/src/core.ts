import type { ESTree } from "meriyah";

import type {
  ExpressionPrecedenceType,
  IOptions,
  CodeGeneratorType,
} from "./types";
import { EXPRESSIONS_PRECEDENCE } from "./constants";
import { GENERATOR } from "./generator";

export class State {
  output: string = "";
  generator: CodeGeneratorType;
  expressionsPrecedence: ExpressionPrecedenceType;
  indent: string;
  lineEnd: string;
  indentLevel: number;
  writeComments: boolean;

  constructor(options: IOptions = {}) {
    this.generator = options.generator != null ? options.generator : GENERATOR;
    this.expressionsPrecedence =
      options.expressionsPrecedence != null
        ? options.expressionsPrecedence
        : EXPRESSIONS_PRECEDENCE;
    // Formating setup
    this.indent = options.indent != null ? options.indent : "  ";
    this.lineEnd = options.lineEnd != null ? options.lineEnd : "\n";
    this.indentLevel =
      options.startingIndentLevel != null ? options.startingIndentLevel : 0;
    this.writeComments = options.comments ? options.comments : false;
  }

  write(code) {
    this.output += code;
  }
}

export function generate(node: ESTree.Program, options?: IOptions) {
  /*
    Returns a string representing the rendered code of the provided AST `node`.
    The `options` are:
  
    - `indent`: string to use for indentation (defaults to `␣␣`)
    - `lineEnd`: string to use for line endings (defaults to `\n`)
    - `startingIndentLevel`: indent level to start from (defaults to `0`)
    - `comments`: generate comments if `true` (defaults to `false`)
    - `output`: output stream to write the rendered code to (defaults to `null`)
    - `generator`: custom code generator (defaults to `GENERATOR`)
    - `expressionsPrecedence`: custom map of node types and their precedence level (defaults to `EXPRESSIONS_PRECEDENCE`)
    */
  const state = new State(options);
  // Travel through the AST node and generate the code
  state.generator[node.type](node, state);
  return state.output;
}
