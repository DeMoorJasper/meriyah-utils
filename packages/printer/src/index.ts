export * from "./core";
export * from "./generator";
export * from "./utils";
export * from "./constants";
