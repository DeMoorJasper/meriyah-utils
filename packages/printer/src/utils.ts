import { ESTree } from "meriyah";

import { State } from "./core";
import { NEEDS_PARENTHESES, OPERATOR_PRECEDENCE } from "./constants";

export function formatSequence(state: State, nodes: Array<ESTree.Node>) {
  // Writes into `state` a sequence of `nodes`.
  const { generator } = state;
  state.write("(");
  if (nodes != null && nodes.length > 0) {
    generator[nodes[0].type](nodes[0], state);
    const { length } = nodes;
    for (let i = 1; i < length; i++) {
      const param = nodes[i];
      state.write(", ");
      generator[param.type](param, state);
    }
  }
  state.write(")");
}

export function expressionNeedsParenthesis(
  state: State,
  node,
  parentNode,
  isRightHand: boolean
) {
  const nodePrecedence = state.expressionsPrecedence[node.type];
  if (nodePrecedence === NEEDS_PARENTHESES) {
    return true;
  }
  const parentNodePrecedence = state.expressionsPrecedence[parentNode.type];
  if (nodePrecedence !== parentNodePrecedence) {
    // Different node types
    return (
      (!isRightHand &&
        nodePrecedence === 15 &&
        parentNodePrecedence === 14 &&
        parentNode.operator === "**") ||
      nodePrecedence < parentNodePrecedence
    );
  }
  if (nodePrecedence !== 13 && nodePrecedence !== 14) {
    // Not a `LogicalExpression` or `BinaryExpression`
    return false;
  }
  if (node.operator === "**" && parentNode.operator === "**") {
    // Exponentiation operator has right-to-left associativity
    return !isRightHand;
  }
  if (isRightHand) {
    // Parenthesis are used if both operators have the same precedence
    return (
      OPERATOR_PRECEDENCE[node.operator] <=
      OPERATOR_PRECEDENCE[parentNode.operator]
    );
  }
  return (
    OPERATOR_PRECEDENCE[node.operator] <
    OPERATOR_PRECEDENCE[parentNode.operator]
  );
}

export function formatExpression(
  state: State,
  node,
  parentNode,
  isRightHand: boolean
) {
  // Writes into `state` the provided `node`, adding parenthesis around if the provided `parentNode` needs it. If `node` is a right-hand argument, the provided `isRightHand` parameter should be `true`.
  const { generator } = state;
  if (expressionNeedsParenthesis(state, node, parentNode, isRightHand)) {
    state.write("(");
    generator[node.type](node, state);
    state.write(")");
  } else {
    generator[node.type](node, state);
  }
}

export function reindent(
  state: State,
  text: string,
  indent: string,
  lineEnd: string
) {
  // Writes into `state` the `text` string reindented with the provided `indent`.
  const lines = text.split("\n");
  const end = lines.length - 1;
  state.write(lines[0].trim());
  if (end > 0) {
    state.write(lineEnd);
    for (let i = 1; i < end; i++) {
      state.write(indent + lines[i].trim() + lineEnd);
    }
    state.write(indent + lines[end].trim());
  }
}

export function formatComments(
  state: State,
  comments,
  indent: string,
  lineEnd: string
) {
  /*
    Writes into `state` the provided list of `comments`, with the given `indent` and `lineEnd` strings.
    Line comments will end with `"\n"` regardless of the value of `lineEnd`.
    Expects to start on a new unindented line.
  */
  const { length } = comments;
  for (let i = 0; i < length; i++) {
    const comment = comments[i];
    state.write(indent);
    if (comment.type[0] === "L") {
      // Line comment
      state.write("// " + comment.value.trim() + "\n");
    } else {
      // Block comment
      state.write("/*");
      reindent(state, comment.value, indent, lineEnd);
      state.write("*/" + lineEnd);
    }
  }
}

export function hasCallExpression(node) {
  // Returns `true` if the provided `node` contains a call expression and `false` otherwise.
  let currentNode = node;
  while (currentNode != null) {
    const { type } = currentNode;
    if (type[0] === "C" && type[1] === "a") {
      // Is CallExpression
      return true;
    } else if (type[0] === "M" && type[1] === "e" && type[2] === "m") {
      // Is MemberExpression
      currentNode = currentNode.object;
    } else {
      return false;
    }
  }
}

export function formatVariableDeclaration(state, node) {
  // Writes into `state` a variable declaration.
  const { generator } = state;
  const { declarations } = node;
  state.write(node.kind + " ");
  const { length } = declarations;
  if (length > 0) {
    generator.VariableDeclarator(declarations[0], state);
    for (let i = 1; i < length; i++) {
      state.write(", ");
      generator.VariableDeclarator(declarations[i], state);
    }
  }
}
