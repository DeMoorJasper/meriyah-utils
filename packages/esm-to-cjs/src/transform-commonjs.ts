import * as meriyah from "meriyah";
import { simpleWalk } from "@meriyah-utils/walker";
import { NodeTypes as n } from "@meriyah-utils/types";
import * as escope from "escope";

import {
  generateAllExportsIterator,
  generateEsModuleSpecifier,
  generateExportGetter,
  generateExportStatement,
} from "./generate-exports";
import {
  generateInteropRequire,
  generateInteropRequireExpression,
  generateRequireStatement,
} from "./generate-require";
import { generateVariableName } from "./utils";

/**
 * Transforms esmodule code into commonjs code, built to be as fast as possible
 */
export function transformCommonJS(program: meriyah.ESTree.Program) {
  const usedVarNames = {};
  const varsToRename = {};
  const trackedExports = {};

  /**
   * All names we export, used to predefine the exports at the start
   */
  const exportNames = new Set<string>();

  const getVarName = (name: string) => {
    let usedName = name.replace(/(\s|\.|-|@|\?|&|=|{|})/g, "");
    while (usedVarNames[usedName]) {
      usedName += "_";
    }
    usedVarNames[usedName] = true;
    return usedName;
  };

  let i = 0;
  let importOffset = 0;

  function addNodeInImportSpace(
    oldPosition: number,
    node: meriyah.ESTree.Statement
  ) {
    program.body.splice(oldPosition, 1);
    program.body.splice(importOffset, 0, node);
    importOffset++;
  }

  let addedSpecifier = false;
  function addEsModuleSpecifier() {
    if (addedSpecifier) {
      return;
    }
    addedSpecifier = true;

    program.body.unshift(generateEsModuleSpecifier());

    // Make sure imports will stay after this
    importOffset++;

    i++;
  }

  let addedDefaultInterop = false;
  function addDefaultInterop() {
    if (addedDefaultInterop) {
      return;
    }
    addedDefaultInterop = true;

    program.body.push(generateInteropRequire());
  }

  /**
   * Adds the export identifiers (exports.a = exports.b = exports.c = void 0)
   */
  function addExportVoids() {
    const exportNamesArray = [...exportNames];
    while (exportNamesArray.length !== 0) {
      // We need to chunk the exports by 50, otherwise this line will get too long
      // and the visitor will create a Maximum call stack size exceeded error
      const exportNamesToUse = exportNamesArray.splice(0, 50);
      const totalNode = {
        type: n.ExpressionStatement,
        expression: {
          type: n.AssignmentExpression,
          operator: "=",
        },
      };
      let currentNode: Partial<meriyah.ESTree.AssignmentExpression> =
        totalNode.expression;
      while (exportNamesToUse.length > 0) {
        const exportName = exportNamesToUse.pop();
        currentNode.left = {
          type: n.MemberExpression,
          object: {
            type: n.Identifier,
            name: "exports",
          },
          property: {
            type: n.Identifier,
            name: exportName,
          } as meriyah.ESTree.Identifier,
        };
        if (exportNamesToUse.length) {
          // @ts-expect-error This will be filled in in the next loop
          currentNode.right = {
            type: n.AssignmentExpression,
            operator: "=",
          } as Partial<meriyah.ESTree.AssignmentExpression>;
          currentNode =
            currentNode.right as Partial<meriyah.ESTree.AssignmentExpression>;
        } else {
          currentNode.right = {
            type: n.UnaryExpression,
            operator: "void",
            prefix: true,
            argument: {
              type: n.Literal,
              value: 0,
            },
          };
        }
      }

      // @ts-expect-error TS thinks this is a partial type, but by now it's full
      program.body.unshift(totalNode);
    }
  }

  // If there is a declaration of `exports` (`var exports = []`), we need to rename this
  // variable as it's a reserved keyword
  let exportsDefined = false;

  simpleWalk(program, (node, parent) => {
    if (node.type === n.VariableDeclaration) {
      // We don't rename exports vars in functions, only on root level
      if (
        parent &&
        parent.type === n.BlockStatement &&
        exportsDefined === false
      ) {
        return true;
      }
    } else if (node.type === n.VariableDeclarator) {
      const declNode = node as meriyah.ESTree.VariableDeclarator;
      if (declNode.id.type === n.Identifier && declNode.id.name === "exports") {
        exportsDefined = true;
      }
    } else if (node.type === n.Identifier && exportsDefined) {
      if (node.name === "exports") {
        node.name = "__$csb_exports";
      }
    } else if (!exportsDefined && parent != null) {
      // Skip, we don't need to go deeper now
      return true;
    }
  });

  for (; i < program.body.length; i++) {
    const statement = program.body[i];

    if (statement.type === n.ExportAllDeclaration) {
      // Handles "export * from './something';"

      addEsModuleSpecifier();
      const { source, exported } = statement;
      if (typeof source.value !== "string") {
        continue;
      }

      const varName = getVarName(`$csb__${generateVariableName(source.value)}`);
      addNodeInImportSpace(i, generateRequireStatement(varName, source.value));
      if (exported) {
        // export * as test from './test';
        // TO:
        // var $csb___test = require("./test");
        // Object.defineProperty(exports, "test", {
        //   enumerable: true,
        //   configurable: true,
        //   get: function $csbGet() {
        //     return $csb___test;
        //   }
        // });

        exportNames.add(exported.name);
        program.body.splice(
          importOffset++,
          0,
          generateExportGetter(
            { type: n.Literal, value: exported.name },
            { type: n.Identifier, name: varName }
          )
        );
      } else {
        // export * from './test';
        // TO:
        // const _csb = require('./test');
        // Object.keys(_csb).forEach(key => {
        //   if (key === 'default' || key === '__esModule')
        //     return;
        //   exports[key] = _csb[key])
        // }

        program.body.push(generateAllExportsIterator(varName));
      }
    } else if (statement.type === n.ExportNamedDeclaration) {
      // export { a } from './test';
      // TO:
      // const _csb = require('./test');
      // exports.a = _csb.a;

      addEsModuleSpecifier();
      if (statement.source) {
        // export { ... } from ''
        const { source } = statement;
        if (typeof source.value !== "string") {
          continue;
        }
        const varName = getVarName(
          `$csb__${generateVariableName(source.value)}`
        );

        if (
          statement.specifiers.length === 1 &&
          statement.specifiers[0].type === n.ExportSpecifier &&
          statement.specifiers[0].local.name === "default"
        ) {
          // In this case there's a default re-export. So we need to wrap it in a interopRequireDefault to make sure
          // that default is exposed.
          addDefaultInterop();
          addNodeInImportSpace(
            i,
            generateInteropRequireExpression(
              {
                type: n.CallExpression,
                callee: {
                  type: n.Identifier,
                  name: "require",
                },
                arguments: [
                  {
                    type: n.Literal,
                    value: source.value,
                  },
                ],
              },
              varName
            )
          );
        } else {
          addNodeInImportSpace(
            i,
            generateRequireStatement(varName, source.value)
          );
        }

        if (statement.specifiers.length) {
          statement.specifiers.forEach((specifier) => {
            if (specifier.type === n.ExportSpecifier) {
              exportNames.add(specifier.exported.name);
              program.body.splice(
                importOffset++,
                0,
                generateExportGetter(
                  { type: n.Literal, value: specifier.exported.name },
                  {
                    type: n.MemberExpression,
                    object: {
                      type: n.Identifier,
                      name: varName,
                    },
                    property: {
                      type: n.Identifier,
                      name: specifier.local.name,
                    },
                  }
                )
              );
            }

            i++;
          });
        }
      } else if (statement.declaration) {
        // First remove the export statement
        program.body[i] = statement.declaration;

        if (
          statement.declaration.type === n.FunctionDeclaration ||
          statement.declaration.type === n.ClassDeclaration
        ) {
          // export function test() {}

          // @ts-ignore
          const varName = statement.declaration.id.name;
          i++;
          // Add to start of the file, after the defineModule for __esModule. This way this export is already
          // defined before requiring other modules. This is only possible for function exports.
          const positionToInsert =
            statement.declaration.type === n.FunctionDeclaration ? 1 : i;
          program.body.splice(
            positionToInsert,
            0,
            generateExportStatement(varName, varName)
          );
          exportNames.add(varName);
        } else {
          // export const a = {}
          const declaration =
            statement.declaration as meriyah.ESTree.VariableDeclaration;

          const exportStatements: meriyah.ESTree.Statement[] = [];
          for (let node of declaration.declarations) {
            if (node.id.type === n.ObjectPattern) {
              // export const { a } = c;
              for (let property of node.id.properties) {
                if (
                  property.type !== n.Property ||
                  property.value.type !== n.Identifier
                ) {
                  continue;
                }

                exportNames.add(property.value.name);
                trackedExports[property.value.name] = property.value.name;
                exportStatements.push(
                  generateExportStatement(
                    property.value.name,
                    property.value.name
                  )
                );
              }
            } else if (node.id.type === n.Identifier) {
              trackedExports[node.id.name] = node.id.name;

              exportNames.add(node.id.name);
              exportStatements.push(
                generateExportStatement(node.id.name, node.id.name)
              );
            } else if (node.id.type === n.ArrayPattern) {
              // export const [a, b] = c;
              for (let property of node.id.elements) {
                if (property.type !== n.Identifier) {
                  continue;
                }

                exportNames.add(property.name);
                trackedExports[property.name] = property.name;
                exportStatements.push(
                  generateExportStatement(property.name, property.name)
                );
              }
            }
          }

          program.body.splice(i, 1, declaration, ...exportStatements);
        }
      } else if (statement.specifiers) {
        program.body.splice(i, 1);
        i--;
        statement.specifiers.forEach((specifier) => {
          if (specifier.type === n.ExportSpecifier) {
            i++;

            exportNames.add(specifier.exported.name);
            program.body.unshift(
              generateExportGetter(
                { type: n.Literal, value: specifier.exported.name },
                { type: n.Identifier, name: specifier.local.name }
              )
            );
            // Make sure that nothing can get inbetween this
            importOffset++;
          }
        });
      }
    } else if (statement.type === n.ExportDefaultDeclaration) {
      addEsModuleSpecifier();
      // export default function() {}
      // export default class A {}
      const varName = getVarName(`$csb__default`);
      // First remove the export statement
      if (statement.declaration) {
        if (statement.declaration.type === n.FunctionDeclaration) {
          // @ts-ignore
          statement.declaration.type = n.FunctionExpression;
        } else if (statement.declaration.type === n.ClassExpression) {
          // @ts-ignore
          statement.declaration.type = n.ClassDeclaration;
        }
        const newDeclaration =
          statement.declaration as meriyah.ESTree.Expression;

        // Create a var with the export
        if (
          statement.declaration.type === n.ClassDeclaration ||
          statement.declaration.type === n.FunctionExpression
        ) {
          if (!statement.declaration.id) {
            // If the function or class has no name, we give it to it
            statement.declaration.id = {
              type: n.Identifier,
              name: varName,
            };
          }

          program.body[i] =
            statement.declaration as meriyah.ESTree.DeclarationStatement;
          i++;

          program.body.splice(
            i,
            0,
            generateExportStatement(statement.declaration.id.name, "default")
          );
        } else {
          program.body[i] = {
            type: n.VariableDeclaration,
            kind: "var" as "var",

            declarations: [
              {
                type: n.VariableDeclarator,
                id: {
                  type: n.Identifier,
                  name: varName,
                },
                init: newDeclaration,
              },
            ],
          };
          i++;

          program.body.splice(
            i,
            0,
            generateExportStatement(varName, "default")
          );
        }

        if (
          newDeclaration.type === n.ClassDeclaration ||
          newDeclaration.type === n.FunctionExpression
        ) {
          // @ts-ignore
          trackedExports[newDeclaration.id.name] = "default";
        }
      }
    } else if (statement.type === n.ImportDeclaration) {
      // @ts-ignore Wrong typing in lib?
      const source: meriyah.ESTree.Literal = statement.source;

      if (typeof source.value !== "string") {
        continue;
      }
      const varName = getVarName(`$csb__${generateVariableName(source.value)}`);

      addNodeInImportSpace(i, generateRequireStatement(varName, source.value));

      statement.specifiers.reverse().forEach((specifier) => {
        let localName: string;
        let importName: string;

        if (specifier.type === n.ImportSpecifier) {
          // import {Test} from 'test';
          // const _test = require('test');
          // var Test = _test.Test;

          // Mark that we need to rename all references to this variable
          // to the new require statement. This will happen in the second pass.
          varsToRename[specifier.local.name] = [
            varName,
            specifier.imported.name,
          ];

          return;
        }
        i++;

        if (specifier.type === n.ImportDefaultSpecifier) {
          // import Test from 'test';
          // const _test = require('test');
          // var Test = interopRequireDefault(_test).default;
          localName = specifier.local.name;
          importName = "default";
          addDefaultInterop();

          program.body.splice(
            // After the require statement
            importOffset,
            0,
            generateInteropRequireExpression(
              { type: n.Identifier, name: varName },
              localName
            )
          );

          varsToRename[localName] = [localName, "default"];
          importOffset++;
          return;
        }

        if (specifier.type === n.ImportNamespaceSpecifier) {
          // import * as Test from 'test';
          // const _test = require('test');
          // var Test = _test;
          localName = specifier.local.name;
          // @ts-ignore
          importName = null;
        }

        // insert in index 1 instead of 0 to be after the interopRequireDefault
        program.body.splice(importOffset, 0, {
          type: n.VariableDeclaration,
          kind: "var" as "var",
          declarations: [
            {
              type: n.VariableDeclarator,
              id: {
                type: n.Identifier,
                // @ts-ignore
                name: localName,
              },
              // @ts-ignore
              init: importName
                ? {
                    type: n.MemberExpression,
                    computed: false,
                    object: {
                      type: n.Identifier,
                      name: varName,
                    },
                    property: {
                      type: n.Identifier,
                      name: importName,
                    },
                  }
                : {
                    type: n.Identifier,
                    name: varName,
                  },
            },
          ],
        });
        importOffset++;
      });
    }
  }

  // console.log(exportNames);
  if (
    Object.keys(varsToRename).length > 0 ||
    Object.keys(trackedExports).length > 0
  ) {
    // Convert all the object shorthands to not shorthands, needed later when we rename variables so we
    // don't change to the key literals
    simpleWalk(program, (node) => {
      if (node.type === n.Property) {
        const property = node as meriyah.ESTree.Property;
        if (
          property.shorthand &&
          property.value.type !== n.AssignmentPattern // Not a default initializer
        ) {
          property.value = {
            ...property.key,
          };
          property.shorthand = false;
        }
      }
    });

    // A second pass where we rename all references to imports that were marked before.
    const scopeManager = escope.analyze(program, {
      ecmaVersion: 6,
      sourceType: "module",
    });

    scopeManager.acquire(program);
    scopeManager.scopes.forEach((scope) => {
      scope.references.forEach((ref) => {
        // If the variable cannot be resolved, it must be the var that we had
        // just changed.
        if (
          Object.prototype.hasOwnProperty.call(
            varsToRename,
            ref.identifier.name
          ) &&
          ref.resolved === null &&
          !ref.writeExpr &&
          ref.from.type !== "class"
        ) {
          ref.identifier.name = `(0, ${varsToRename[ref.identifier.name].join(
            "."
          )})`;
        }

        if (
          Object.prototype.hasOwnProperty.call(
            trackedExports,
            ref.identifier.name
          ) &&
          ref.isWrite() &&
          ref.resolved === null &&
          !ref.init
        ) {
          const name = trackedExports[ref.identifier.name];
          ref.identifier.name = `exports.${name} = ${ref.identifier.name}`;
        }
      });
    });
    scopeManager.detach();
  }

  addExportVoids();
}
