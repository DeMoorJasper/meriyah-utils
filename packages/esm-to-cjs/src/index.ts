import * as meriyah from "meriyah";
import { transformCommonJS } from "./transform-commonjs";

export function convertEsModule(program: meriyah.ESTree.Program): void {
  transformCommonJS(program);
}
