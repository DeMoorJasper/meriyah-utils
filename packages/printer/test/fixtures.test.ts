import fs from "fs";
import path from "path";
import * as meriyah from "meriyah";
import * as globby from "globby";

import { generate } from "../src/index";

function parseCode(code: string): meriyah.ESTree.Program {
  return meriyah.parseModule(code, {
    module: true,
    webcompat: true,
    directives: false,
    next: true,
    raw: true,
    jsx: true,
    loc: false,
    ranges: false,
  });
}

function testFixture(filepath: string, skipInvalid: boolean = false) {
  const input = fs.readFileSync(filepath, "utf-8");
  let inputAST;
  try {
    inputAST = parseCode(input);
  } catch (err) {
    if (skipInvalid) return;
    throw err;
  }

  const output = generate(inputAST);
  const outputAST = parseCode(output);

  try {
    expect(outputAST).toEqual(inputAST);
  } catch (err) {
    console.error("=== INPUT ===\n" + input + "\n\n=== OUTPUT ===\n" + output);
    throw err;
  }
}

describe("Fixtures", () => {
  const FIXTURES_DIR = path.join(__dirname, "fixtures");
  const fixtures = globby.sync("**.js", {
    cwd: FIXTURES_DIR,
  });

  for (let fixture of fixtures) {
    it(`Fixtures: ${fixture}`, () => {
      const filepath = path.join(FIXTURES_DIR, fixture);
      testFixture(filepath);
    });
  }
});

describe("Printer transforms", () => {
  const TRANSFORM_FIXTURES = path.join(__dirname, "transforms");

  // Safari does not support private class properties yet
  it("Should transform private class properties and functions into regular class properties", () => {
    const filepath = path.join(TRANSFORM_FIXTURES, "private-class-props.js");
    const code = fs.readFileSync(filepath, "utf-8");
    const program = parseCode(code);
    const output = generate(program);
    expect(output).toMatchSnapshot();
  });
});

// describe("test262-parser-tests", () => {
//   // We don't support encoded variable names yet
//   const ENCODED_VARIABLE_NAMES = new Set([
//     "596746323492fbfd.js",
//     "5c3d125ce5f032aa.js",
//     "c85fbdb8c97e0534.js",
//     "dafb7abe5b9b44f5.js",
//     "eaee2c64dfc46b6a.js",
//     "f5b89028dfa29f27.js",
//     "f7f611e6fdb5b9fc.js",
//   ]);
//   const TEST_262_MODULE_DIR = path.dirname(
//     require.resolve("test262-parser-tests/package.json")
//   );
//   const TEST_262_DIR = path.join(TEST_262_MODULE_DIR, "pass");
//   const fixtures = globby
//     .sync("**.js", {
//       cwd: TEST_262_DIR,
//     })
//     .filter((f) => !ENCODED_VARIABLE_NAMES.has(f));

//   for (let fixture of fixtures) {
//     it(`test262: ${fixture}`, () => {
//       const filepath = path.join(TEST_262_DIR, fixture);
//       testFixture(filepath, true);
//     });
//   }
// });
