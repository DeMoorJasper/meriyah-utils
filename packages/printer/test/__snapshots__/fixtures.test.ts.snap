// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Printer transforms Should transform private class properties and functions into regular class properties 1`] = `
"class SharedObserver {
  sharedResizeObserver;

  callbacks = new WeakMap();

  static something;

  fn = () => {};

  constructor() {
    const ResizeObserver = safeWindow.ResizeObserver ?? ResizeObserverPolyfill;
    this.sharedResizeObserver = new ResizeObserver(this.updateResizedElements.bind(this));
  }
  test() {
    this.sharedResizeObserver = new ResizeObserver(this.updateResizedElements.bind(this));
    return true;
  }
  privateMethod() {
    return 'hello world';
  }
}
"
`;
